### PR TITLE
Disable sort when limit is explicit

### PR DESCRIPTION
**Pull request by samdark**
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ❌

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -28,3 +28,4 @@ phpunit.phar
 /phpunit.xml
 # phpunit cache
 /.phpunit.cache/*
+.phpunit.result.cache
```
Comment (vjik) [inline .gitignore:31]:
```suggestion
```

Not need for PhpUnit 10+

File: src/Paginator/KeysetPaginator.php
Comment (vjik) [inline src/Paginator/KeysetPaginator.php:1]: Keyset paginator doesn't take into account limit in data reader. Seems, it's not requires changes.

File: src/Paginator/OffsetPaginator.php
Comment (vjik) [inline src/Paginator/OffsetPaginator.php:1]: Add line to changelog.

File: src/Paginator/OffsetPaginator.php [hunk 2]
```diff
@@ -237,6 +241,10 @@ public function withSort(?Sort $sort): static
 
     public function getSort(): ?Sort
     {
+        if ($this->dataReader instanceof LimitableDataInterface && $this->dataReader->getLimit() !== null) {
```
Comment (vjik) [inline src/Paginator/OffsetPaginator.php:244]: I think, it's not need, because get current sort for non-sortable data reader (with limit) is OK.

**Review by vjik — CHANGES_REQUESTED** [review]

File: src/Paginator/KeysetPaginator.php [hunk 3]
```diff
@@ -269,6 +269,10 @@ public function withSort(?Sort $sort): static
 
     public function getSort(): ?Sort
     {
+        if ($this->dataReader instanceof LimitableDataInterface && $this->dataReader->getLimit() !== null) {
+            return null;
+        }
```
Comment (samdark) [inline src/Paginator/KeysetPaginator.php:274]: Keyset paginator can not work without sorting. Does that mean that limit can't be applied at all?

File: src/Paginator/KeysetPaginator.php [hunk 4]
```diff
@@ -257,7 +257,7 @@ public function getNextToken(): ?PageToken
 
     public function isSortable(): bool
     {
-        return true;
+        return !($this->dataReader instanceof LimitableDataInterface && $this->dataReader->getLimit() !== null);
```
Comment (samdark) [inline src/Paginator/KeysetPaginator.php:260]: Keyset paginator can't be non-sortable, right @vjik ?

Comment (vjik) [reply src/Paginator/KeysetPaginator.php:260]: Yes.

File: tests/Paginator/OffsetPaginatorTest.php [hunk 5]
```diff
@@ -561,7 +560,7 @@ public function testWithSortNonSortableData(): void
     {
         $paginator = new OffsetPaginator(new StubOffsetData());
 
-        $this->expectException(LogicException::class);
+        $this->expectException(InvalidArgumentException::class);
```
Comment (samdark) [inline tests/Paginator/OffsetPaginatorTest.php:563]: Not sure it worth changing exception type.

**Comment by samdark** [comment]
Applied suggestions and fixed some additional stuff.

File: src/Paginator/OffsetPaginator.php [hunk 6]
```diff
@@ -226,16 +225,21 @@ public function getTotalPages(): int
...
+
         return $this->dataReader instanceof SortableDataInterface;
     }
 
     public function withSort(?Sort $sort): static
     {
-        if (!$this->dataReader instanceof SortableDataInterface) {
-            throw new LogicException('Data reader does not support sorting.');
+        if (!$this->isSortable()) {
+            throw new InvalidArgumentException('Data reader does not support sorting.');
```
Comment (vjik) [inline src/Paginator/OffsetPaginator.php:238]: Argument is correct. Suggest keep `LogicException`, it's notifying the user to fix his code.

File: src/Paginator/OffsetPaginator.php [hunk 7]
```diff
@@ -226,16 +225,21 @@ public function getTotalPages(): int
...
         }
 
         $new = clone $this;
+        /** @psalm-suppress UndefinedInterfaceMethod */
```
Comment (vjik) [inline src/Paginator/OffsetPaginator.php:242]:
```suggestion
```

Add `@psalm-assert-if-true SortableDataInterface $this->dataReader` annotation to `isSortable()` method.

File: src/Paginator/OffsetPaginator.php [hunk 8]
```diff
@@ -252,11 +256,12 @@ public function isFilterable(): bool
...
     public function withFilter(FilterInterface $filter): static
     {
-        if (!$this->dataReader instanceof FilterableDataInterface) {
-            throw new LogicException('Data reader does not support filtering.');
+        if (!$this->isFilterable()) {
+            throw new InvalidArgumentException('Data reader does not support filtering.');
         }
 
         $new = clone $this;
+        /** @psalm-suppress UndefinedInterfaceMethod */
```
Comment (vjik) [inline src/Paginator/OffsetPaginator.php:264]:
```suggestion
```

Add `@psalm-assert-if-true FilterableDataInterface $this->dataReader` annotation to `isFiltarable()` method.

File: src/Paginator/OffsetPaginator.php [hunk 9]
```diff
@@ -252,11 +256,12 @@ public function isFilterable(): bool
 
     public function withFilter(FilterInterface $filter): static
     {
-        if (!$this->dataReader instanceof FilterableDataInterface) {
-            throw new LogicException('Data reader does not support filtering.');
+        if (!$this->isFilterable()) {
+            throw new InvalidArgumentException('Data reader does not support filtering.');
```
Comment (vjik) [inline src/Paginator/OffsetPaginator.php:260]: Argument is correct, but logic of paginator usage in user code is broken. Suggest keep `LogicException`, it's notifying the user to fix his code.

**Review by vjik — CHANGES_REQUESTED** [review]
What about similar fixes in `KeysetPaginator`?

**Comment by samdark** [comment]
@vjik good suggestions. Applied.

**Review by vjik — APPROVED** [review]

**Pull request merged**